model/reliability/query.go

```diff
@@ -1,6 +1,181 @@
+	Distros       []string
+
+	GroupNumDays int
+	GroupBy      stats.GroupBy
+	StartAt      *stats.StartAt
+	Limit        int
+	Sort         stats.Sort
+	Significance float64
+}
+
+// ValidateForTaskReliability validates that the StartAt struct is valid for use with test stats.
+func (f *TaskReliabilityFilter) ValidateForTaskReliability() error {
+	catcher := grip.NewBasicCatcher()
+	statsFilter := stats.StatsFilter{
+		Project:       f.Project,
+		Requesters:    f.Requesters,
+		AfterDate:     f.AfterDate,
+		BeforeDate:    f.BeforeDate,
+		Tasks:         f.Tasks,
+		BuildVariants: f.BuildVariants,
+		Distros:       f.Distros,
+		GroupNumDays:  f.GroupNumDays,
+		GroupBy:       f.GroupBy,
+		StartAt:       f.StartAt,
+		Limit:         f.Limit,
+		Sort:          f.Sort,
+	}
+
+	catcher.Add(statsFilter.ValidateCommon())
+
+	if f.Limit > MaxQueryLimit || f.Limit <= 0 {
+		catcher.Add(errors.New("Invalid Limit value"))
+	}
+
+	if f.Significance > MaxSignificanceLimit || f.Significance < MinSignificanceLimit {
+		catcher.Add(errors.New("Invalid Significance value"))
+	}
+
+	if len(f.Tasks) == 0 {
+		catcher.Add(errors.New("Missing Task values"))
+	}
+	return catcher.Resolve()
+}
+
+//////////////////////////////
+// Task Reliability Querying //
+//////////////////////////////
+
+// TaskReliability represents task execution statistics.
+type TaskReliability struct {
+	TaskName     string    `bson:"task_name"`
+	BuildVariant string    `bson:"variant"`
+	Distro       string    `bson:"distro"`
+	Date         time.Time `bson:"date"`
+
+	NumTotal           int     `bson:"num_total"`
+	NumSuccess         int     `bson:"num_success"`
+	NumFailed          int     `bson:"num_failed"`
+	NumTimeout         int     `bson:"num_timeout"`
+	NumTestFailed      int     `bson:"num_test_failed"`
+	NumSystemFailed    int     `bson:"num_system_failed"`
+	NumSetupFailed     int     `bson:"num_setup_failed"`
+	AvgDurationSuccess float64 `bson:"avg_duration_success"`
+	SuccessRate        float64
+	Z                  float64
+	LastUpdate         time.Time `bson:"last_update"`
+}
+
+// calculateSuccessRate using
+// // https://en.wikipedia.org/wiki/Binomial_proportion_confidence_interval#Wilson_score_interval
+// and return the lower value (for success rates).
+func (s *TaskReliability) calculateSuccessRate() {
+	total := float64(s.NumTotal)
+	success := float64(s.NumSuccess)
+	failed := float64(s.NumFailed)
+	low := 0.0
+	high := 0.0
+	p := 0.0
+
+	if total != 0 {
+		product := success * failed
+		p = success / total
+
+		zSquared := s.Z * s.Z
+		zSquared2 := zSquared / 2.0
+		zSquared4 := zSquared / 4.0
+
+		value := (success + zSquared2) / total / (1 + zSquared/total)
+		margin := (s.Z * math.Sqrt(product/total+zSquared4) / total) / (1 + zSquared/total)
+		high = value + margin
+		low = value - margin
+
+	}
+	grip.Debugf("SuccessRate: %.4f(p=%.4f,hi=%.4f). Success=%d, Failed=%d, Total=%d, z=%4f\n", low, p, high, int(success), int(failed), int(total), s.Z)
```

Comment:
How useful will this logging be? If you're going to have it here, you should use message.fields so that it's easier to search for